src/control/faults/faults.go

```diff
@@ -0,0 +1,113 @@
+package faults
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+// Resolution represents a potential fault resolution.
+type Resolution string
+
+const (
+	// ResolutionEmpty is equivalent to an empty string.
+	ResolutionEmpty = Resolution("")
+	// ResolutionUnknown indicates that there is no known
+	// resolution for the fault.
+	ResolutionUnknown = Resolution("no known resolution")
+	// ResolutionNone indicates that the fault cannot be
+	// resolved.
+	ResolutionNone = Resolution("none")
+)
+
+func (r Resolution) String() string {
+	return string(r)
+}
+
+const (
+	UnknownDomainStr      = "unknown"
+	UnknownDescriptionStr = "unknown fault"
+)
+
+var (
+	// UnknownFault represents an unknown fault.
+	UnknownFault = &Fault{
+		Code:       CodeUnknown,
+		Resolution: ResolutionUnknown,
+	}
+)
+
+// Fault represents a well-known error specific to a domain,
+// along with an optional potential resolution for the error.
+//
+// It implements the error interface and can be used
+// interchangeably with regular "dumb" errors.
+type Fault struct {
+	Domain      string
```

Comment:
Not a change I'd insist on, but "domain" is slightly confusing terminology here. Took me a bit of reading the code to figure out it is just an arbitrary prefix for the fault string.

Reply:
It's a fairly common term in the context of formal error handling. Fault/Failure Domain indicates a logical grouping of faults (e.g. storage, network, security, etc). I'm open to alternatives which might be less confusing though.

Reply:
Even just calling it a "fault domain" in the comment would be enough to clarify.

Reply:
Done.